format/view-spec.md

```diff
@@ -97,7 +97,10 @@ Summary is a string to string map of metadata about a view version. Common metad
 
 View definitions can be represented in multiple ways. Representations are documented ways to express a view definition.
 
-A view version can have more than one representation. All representations for a version must express the same underlying definition. Engines are free to choose the representation to use.
+A view version can have more than one representation. All representations for a version must express the same underlying definition. Engines are free to choose the representation to use. For the table identifiers in the view definitions to be interoperable across engines, the following must be true:
+* All engines must resolve a fully specified SQL identifier to the same table in the same catalog.
```

Comment:
Could I suggest this?

- When table identifier is fully specified, then engines should resolve to the same catalog and table.
- When table identifier omits the catalog name, then engines should use the default-catalog or the view's catalog (which is allowed to be engine specific) and resolve to the same catalog and table.
- When table identifier omits the namespace, then engines should use the default-namespace to resolve to the same catalog and table.

Reply:
+1 for Benny's suggestion.

Reply:
Nit: I think putting 'all' in front of engines in each sentence will make it clearer, ie,
```When table identifier is fully specified, then all engines should resolve to the same catalog and table.```

etc

Reply:
> then engines should use the default-catalog or the view's catalog (which is allowed to be engine specific) and resolve to the same catalog and table.

I think use `x or y` will weaken the contract. Typically default catalog is used when the catalog is omitted.

Reply:
> (which is allowed to be engine specific)

Not clear what that means.

Reply:
Apart from the above, the main point is that Benny's recommendation does not address the core issue this PR is trying to address, which is "sharing" both the default namespace and the default catalog names across engines. This will make partially specified identifiers portable across engines. Just "using the default namespace" does not guarantee portability.

Let’s consider two engines, Engine A and Engine B and the following shared view SQL:

```sql
SELECT * FROM y.z;
```

Here, `y.z` is a partially specified identifier, with the catalog name omitted.

Under Benny's version of the spec:

* Engine A uses its default catalog (`x1`) and resolves the identifier as `x1.y.z`.
* Engine B uses its default catalog (`x2`) and resolves the identifier as `x2.y.z`.
* Both engines follow the suggested spec by "using the default catalog" to resolve the identifier, but they resolve it to different catalogs and tables because their default catalog names are not shared.

Reply:
> Typically default catalog is used when the catalog is omitted.
When catalog is omitted, it also could be the case of using the same catalog where view is defined.

@wmoustafa  I feel here is the key part from Benny's suggestion
```
resolve to the same catalog and table
```

Benny's wording reads a little more clear to me because it specify the expected behavior under `when` condition.

Reply:
Two points about the new suggestion:
* It still has the gap in my first two comments [here](https://github.com/apache/iceberg/pull/11365/files#r1862650411) and [here](https://github.com/apache/iceberg/pull/11365/files#r1862650874).
* It makes the sharing of the catalog/namespace dependent on `when catalog/namespace is omitted`. There is no dependency here and those behaviors/conditions should be exhibited by the engines regardless of the how individual views or identifiers are written. So I see this format is more confusing.

Reply:
There are two parts to consider: 
1. Resolution of table references
3. Portability constraints

I think @bennychow 's suggestion conflates the two. In addition to the portability constraint, it describes how a query engine should resolve a partial table reference using either the default-catalog or the view catalog. This conflicts with how table references are currently resolved in the view spec. Currently only the default-catalog is used.

If the main point of this PR is to define the Portability constraints, then we should not change the resolution of table references. Changing the table reference resolution would also be a breaking change to the view spec.

Reply:
So, I was imagining that table resolution in the query engine would work like this:

1. Assume SQL identifier is already fully qualified and so first segment is the catalog name.  Query engine would see if this catalog exists and can resolve the remaining segments into a table/view.
2. Prepend the view's catalog and default-namespace (optional) to the SQL identifier.  Now, try to resolve with the query engine's registered catalogs.
3. If default-catalog is set, prepend the default-catalog and default-namespace (optional) to the SQL identifier.  Again, try to resolve with query engine's registered catalogs.

I don't think there is anything new with item 1 and 3 but item 2 is interesting because it allows two different engines to reference the same table using different catalog names.  This works only when the table and view are stored in the same catalog.

Reply:
> then we should not change the resolution of table references. Changing the table reference resolution would also be a breaking change to the view spec.

@JanKaul Could you clarify what you meant by this, preferably through an example?

Reply:
I would assume the resolution of the identifiers to work something like this:

0. Get list of available catalogs
1. Set the table-name equal to the last element.
2. Check if the second to last element exists.
 - If it does, set the namespace equal to it.
 - If it doesn't, use the default-namespace
3. Check if the next element in reverse direction exists
 - if the element is equal to an available catalog, set catalog-name to it
 - if the element exists and is not an available catalog, prepend it to the namespace and continue with step 3.
 - if the element doesn't exist, use the default-catalog 
4. Check if the resolved catalog, namespace, and table exist

The important thing is that if the namespace or catalog are not present, the default-namespace and default-catalog from the view version are used. The resolution currently doesn't fallback to use the view catalog. 

If we would now opt to use the view catalog instead of the default-catalog if the catalog is not present, the behavior would be different compared to the current view spec. This is what I mean with breaking change.

Reply:
There are two ways to approach this:
* Deprecate using view catalog for table identifier resolution.
* Or assume that the above (i.e., guidance from this PR) only kicks in if view catalog is null.

I am in favor of the former despite being a breaking change. It is simpler, and compatible with how engines with views work generally. There is a thread [here](https://docs.google.com/document/d/1e5orD_sBv0VlNNLZRgUtalVUllGuztnAGTtqo8J0UG8/edit?disco=AAABWhyI1tQ) with the same sentiment from @ajantha-bhat.

Reply:
If these are just recommendations should we just say "Identifiers should always contain namespace information.
Catalog name should be omitted from any identifiers. Catalog names are engine independent so view representations should omit the catalog name when writing view representations. When reading engines should use the catalog name of the View Catalog as the catalog name when resolving identifiers."

Reply:
@RussellSpitzer I think recommending the catalog part to be omitted is not very practical. Additionally, I am not in favor of "recommending" in a spec, because it leaves the door open about what is really supported vs not.

That said, these are actually not intended to be "recommendations" on how the views are expressed, but rather "expectations/assumptions" about the multi-engine environment. Once those expectations are met, users can safely use any of `catalog.db.table`, `db.table`, or `table` without worrying about cross engine inconsistency. This is the objective of the PR. More details in [this doc](https://docs.google.com/document/d/1e5orD_sBv0VlNNLZRgUtalVUllGuztnAGTtqo8J0UG8/edit?tab=t.0) (specifically up to -and including- the "Portable SQL table identifiers" section).

Reply:
How does a user know what "catalog" should be set to in their engine config? That's the only thing i'm worried about. There is basically only ever going to be one correct answer for every table in the Iceberg Catalog so I don't like persisting in every entry.

Reply:
The goal of this PR is to clarify what is needed/recommended for interoperability across engines.
- if catalog name is not specified, default catalog is used. if default catalog is not defined, then view catalog is used.
- if catalog name is specified, it should be the same across engines. 

It is still valid to define a view without considering engine interoperability.